cmd/osv-scanner/fix/noninteractive.go

```diff
@@ -247,6 +247,113 @@ func relockUnfixableVulns(diffs []resolution.ResolutionDiff) []*resolution.Resol
+	}
+
+	r.Infof("Resolving %s...\n", opts.Manifest)
+	f, err := lockfile.OpenLocalDepFile(opts.Manifest)
+	if err != nil {
+		return err
+	}
+
+	manif, err := opts.ManifestRW.Read(f)
+	f.Close()
+	if err != nil {
+		return err
+	}
+
+	opts.Client.PreFetch(ctx, manif.Requirements, manif.FilePath)
+	res, err := resolution.Resolve(ctx, opts.Client, manif)
+	if err != nil {
+		return err
+	}
+
+	if errs := res.Errors(); len(errs) > 0 {
+		r.Warnf("WARNING: encountered %d errors during dependency resolution:\n", len(errs))
+		r.Warnf(resolutionErrorString(res, errs))
+	}
+
+	res.FilterVulns(opts.MatchVuln)
+	// TODO: count vulnerabilities per unique version as scan action does
+	totalVulns := len(res.Vulns)
+	r.Infof("Found %d vulnerabilities matching the filter\n", totalVulns)
+
+	allPatches, err := remediation.ComputeOverridePatches(ctx, opts.Client, res, opts.RemediationOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := opts.Client.WriteCache(manif.FilePath); err != nil {
+		r.Warnf("WARNING: failed to write resolution cache: %v\n", err)
+	}
+
+	if len(allPatches) == 0 {
+		r.Infof("No dependency patches are possible\n")
+		r.Infof("REMAINING-VULNS: %d\n", totalVulns)
+		r.Infof("UNFIXABLE-VULNS: %d\n", totalVulns)
+
+		return nil
+	}
+
+	depPatches, nFixed := autoChooseOverridePatches(allPatches, maxUpgrades)
+	nUnfixable := len(relockUnfixableVulns(allPatches))
+	r.Infof("Can fix %d/%d matching vulnerabilities by overriding %d dependencies\n", nFixed, totalVulns, len(depPatches))
+	for _, p := range depPatches {
+		r.Infof("OVERRIDE-PACKAGE: %s,%s\n", p.Pkg.Name, p.NewRequire)
+	}
+	r.Infof("REMAINING-VULNS: %d\n", totalVulns-nFixed)
+	r.Infof("UNFIXABLE-VULNS: %d\n", nUnfixable)
+	// TODO: Print the FIXED-VULN-IDS, REMAINING-VULN-IDS, UNFIXABLE-VULN-IDS
+	// TODO: Consider potentially introduced vulnerabilities
+
+	r.Infof("Rewriting %s...\n", opts.Manifest)
+	if err := manifest.Overwrite(opts.ManifestRW, opts.Manifest, manifest.ManifestPatch{Manifest: &manif, Deps: depPatches}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func autoChooseOverridePatches(diffs []resolution.ResolutionDiff, maxUpgrades int) ([]manifest.DependencyPatch, int) {
+	if maxUpgrades == 0 {
+		return nil, 0
+	}
+
+	var patches []manifest.DependencyPatch
+	pkgChanged := make(map[resolve.PackageKey]bool) // dependencies we've already applied a patch to
+	fixedVulns := make(map[string]bool)             // vulns that have already been fixed by a patch
+	for _, diff := range diffs {
+		// If this patch is incompatible with existing patches, skip adding it to the patch list.
+
+		// A patch is incompatible if any of its changed packages have already been changed by an existing patch.
+		if slices.ContainsFunc(diff.Deps, func(dp manifest.DependencyPatch) bool { return pkgChanged[dp.Pkg] }) {
+			continue
+		}
+		// A patch is also incompatible if any fixed vulnerability has already been fixed by another patch.
```

Comment:
can you explain a bit more why this makes patches incompatible?

Reply:
Expanded more with an example.